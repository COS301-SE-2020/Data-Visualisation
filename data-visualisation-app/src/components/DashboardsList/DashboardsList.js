/**
 *   @file Dashboards.js
 *   Project: Data Visualisation Generator
 *   Copyright: Open Source
 *   Organisation: Doofenshmirtz Evil Incorporated
 *
 *   Update History:
 *   Date        Author              Changes
 *   -------------------------------------------------------
 *   1/7/2020    Gian Uys      Original
 *
 *   Error Messages: "Error"
 *   Assumptions: None
 *   Constraints: None
 */

import React, {useEffect, useState} from 'react';
import '../../globals/globals.scss';
import './DashboardsList.scss';
import {Typography} from 'antd';
import request from '../../globals/requests';
import * as constants from '../../globals/constants';

const { Title, Paragraph } = Typography;

/**
 *   @class HomePanelButton
 *   @brief Component to display the dashboard name and description.
 */
function HomePanelButton(props) {
    const sizeStyles = {
        width: 300,
        height: 300,
    };

    function getContentPositionStyle() {
        return {
            marginTop: sizeStyles.height / 5 - 10,
        };
    }
    function getContentPositionStyle1() {
        return {
            marginTop: sizeStyles.height / 16,
        };
    }

    function getSizeStyle() {
        return {
            width: sizeStyles.width + 'px',
            height: sizeStyles.height + 'px',
        };
    }
    function comp() {
        if (props.isAddButton) {
            return (
                <div
                    className='dashboardList__container--item'
                    style={{paddingTop: '100px', paddingLeft: '100px'}}
                    onClick={() => props.action()}>
                    <div className='panelLayout home-panel-add' id='button__addDashboards'>+</div>
                </div>
            );
        } else {
            return (
                <div
                    className='dashboardList__container--item panelLayout panelStyling'
                    style={props.colour}
                    onClick={() => props.action()}>
                        <Title ellipsis level={(props.data.name.length < 14 ? 2 : (props.data.name.length > 20 ? 4 : 3))} style={{...getContentPositionStyle(), color: 'white'}}>{props.data.name}</Title>
                        <Paragraph ellipsis={{rows: 7}} style={ {...getContentPositionStyle1(), color: 'white'}} >{props.data.description}</Paragraph>
                </div>
            );
        }
    }

    return <div>{comp()}</div>;
}

/**
 *   @class DashboardsList
 *   @brief Component to display the the list of dashboards.
 *   @details Displays every dashboard name and description in a block grid.
 */
function DashboardsList(props) {

    const [isReady, setIsReady] = useState(false);
    const [dashboardList, setDashboardList] = useState([]);

    useEffect(() => {


        if (request.user.isLoggedIn) {
            request.dashboard.list(function(result) {
                setDashboardList(request.cache.dashboard.list.data);
                setIsReady(true);
            });
        } else {
            // todo: handle when user is not logged in
        }


    }, []);


    const backgrounds = [{
        background: 'linear-gradient(16deg, rgba(85, 95, 233,1) 0%, rgba(85, 95, 233,1) 79%)'
    }, {
        background: 'linear-gradient(16deg, rgba(85, 95, 233,1) 0%, rgba(85, 95, 233,1) 79%)'
    }, {
        background: 'linear-gradient(16deg, rgba(85, 95, 233,1) 0%, rgba(85, 95, 233,1) 79%)'
    }, {
        background: 'linear-gradient(16deg, rgba(85, 95, 233,1) 0%, rgba(85, 95, 233,1) 79%)'
    }, {
        background: 'linear-gradient(16deg, rgba(85, 95, 233,1) 0%, rgba(85, 95, 233,1) 79%)'
    }, {
        background: 'linear-gradient(16deg, rgba(85, 95, 233,1) 0%, rgba(85, 95, 233,1) 79%)'
    }];

    let colorStack = [];
    for (let c = 0; c < backgrounds.length; c++) {
        colorStack.push(c);
    }

    return (
        <div className='content--padding' id='dashboardList__container'>

            {isReady ?
                <React.Fragment>
                    {(() => {
                            return dashboardList.map((dashboard) => {
                                return (
                                    <HomePanelButton
                                        colour={backgrounds[(colorStack.length > 0 ? colorStack.pop() : Math.floor(Math.random() * Math.floor(backgrounds.length)))]}
                                        data={dashboard}
                                        key={dashboard.id}
                                        isAddButton={false}
                                        action={() => {props.setDetails(dashboard.name, dashboard.description); props.setDashboardIndex(dashboard.id);}}
                                    />
                                );
                            });
                        }
                    )()}
                    <HomePanelButton
                        colour={backgrounds[(colorStack.length > 0 ? colorStack.pop() : Math.floor(Math.random() * Math.floor(backgrounds.length)))]}
                        isAddButton={true}
                        action={() => props.onAddButtonClick(true)}
                    />
                </React.Fragment>
            :
                <div style={{textAlign: 'center'}}>
                    {constants.LOADER}
                </div>
            }

        </div>
    );
}

export default DashboardsList;
