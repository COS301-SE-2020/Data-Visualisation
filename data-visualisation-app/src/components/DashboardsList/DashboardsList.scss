.panelLayout {
    margin: 20px;
    padding: 20px;
    float: left;
    overflow: hidden;
}

.home-panel-add {
    border: 1px dashed;
    border-color: #777777;
    border-radius: 3%;
    font-size: 60px;
    color: #b0b0b0;
    //text-align: center;
    //padding-top: -20px;
    //width: 100px;
    transition: border-color, color 0.4s linear;
    //transition: color 0.4s linear;
}

.home-panel-add:hover {
    border-color: #343434;
    color: #343434;
}

.panelStyling {
    /*border: 1px solid #e8e8e8;*/
    border-radius: 2%;
    background: #D83864;
    color: white;
    box-shadow: 0 0px 20px #b4b4b4;

    transition: box-shadow 0.3s ease-in-out;
    animation-play-state: paused;
}

.panelStyling:hover{
    box-shadow: 0 0px 30px #848484;
}

.closeSize {
    width: 12px;
    height: 12px;
    margin: 10px;
}
.closeButton {
    float: right;
}

#button__addDashboards {
    width: 100px;
    height: 100px;
    text-align: center;
    vertical-align: middle;
    line-height: 40px;
}