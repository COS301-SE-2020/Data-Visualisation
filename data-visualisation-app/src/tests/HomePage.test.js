import React from 'react';
import DashboardsList from '../components/DashboardsList';
import renderer from 'react-test-renderer';


Object.defineProperty(window, 'matchMedia', {
    writable: true,
    value: jest.fn().mockImplementation(query => ({
        matches: false,
        media: query,
        onchange: null,
        addListener: jest.fn(), // deprecated
        removeListener: jest.fn(), // deprecated
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn(),
    })),
});

describe('DashboardsList component', () => {
    it('DashboardsList renders correctly', () => {
        const rendered = renderer.create(<DashboardsList/>);
        expect(rendered.toJSON()).toMatchSnapshot();
    });
});

