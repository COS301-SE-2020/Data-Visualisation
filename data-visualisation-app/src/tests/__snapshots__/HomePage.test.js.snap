// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage component HomePage renders correctly 1`] = `
<div
  className="content--padding"
>
  <div
    style={
      Object {
        "textAlign": "center",
      }
    }
  >
    <div
      className="ant-spin ant-spin-spinning"
    >
      <span
        className="ant-spin-dot ant-spin-dot-spin"
      >
        <i
          className="ant-spin-dot-item"
        />
        <i
          className="ant-spin-dot-item"
        />
        <i
          className="ant-spin-dot-item"
        />
        <i
          className="ant-spin-dot-item"
        />
      </span>
    </div>
  </div>
</div>
`;
